### PR TITLE
GFORMS-383 - Introduce IncludeIf functionality at field level

### PR DESCRIPTION
**Pull request by josef-vlach**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala [hunk 1]
```diff
@@ -612,48 +610,58 @@ class SectionRenderingService(frontendAppConfig: FrontendAppConfig, lookupRegist
...
-              validationResult,
-              ei,
-              obligations)
-          case FileUpload() =>
-            htmlForFileUpload(formComponent, formTemplateId, ei, ei.retrievals, validationResult)
-          case InformationMessage(infoType, infoText) =>
-            htmlForInformationMessage(formComponent, infoType, infoText, ei)
-          case htp @ HmrcTaxPeriod(idType, idNumber, regimeType) =>
-            htmlForHmrcTaxPeriod(formComponent, ei, validationResult, obligations, htp)
+        val isVisible = formComponent.includeIf.fold(true) { includeIf =>
```
Comment (girishkamat) [inline app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala:613]: why is the evaluation needed here? By the time we are here, the renderPageOptics would have already filtered invisible form components based on the includeIf evaluation right?

File: test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        val visibilityOptics: FormModelVisibilityOptics[DataOrigin.Mongo] =
+          fmb.visibilityModel[DataOrigin.Mongo, SectionSelectorType.Normal](data, None)
+
+        val formModelOptics: FormModelOptics[DataOrigin.Mongo] =
+          fmb.renderPageModel[DataOrigin.Mongo, SectionSelectorType.Normal](visibilityOptics, None)
+
+        val expected: FormModel[Visibility] = fromPagesWithIndex(expectedPages, staticTypeInfo)
+
+        formModelOptics.formModelVisibilityOptics.formModel shouldBe expected
+        visibilityOptics.recData.variadicFormData shouldBe expectedData
```
Comment (girishkamat) [inline test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala:93]: do we really need the second assertion here?

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala (hunk 1 above)
Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala:613]: No, `renderPageOptics` contains all the components. It is `formModelVisibilityOptics` which contain only visible ones.

The reason `renderPageOptics` contains all components, is that we need to know what components are on the page, so they get rendered accordingly.

File: test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala (hunk 2 above)
Comment (josef-vlach) [reply test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala:93]: It is there as a sanity check that invisibility of the component doesn't affect data we store for such a component.

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala (hunk 1 above)
Comment (girishkamat) [reply app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala:613]: Got it, thanks

File: test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala (hunk 2 above)
Comment (girishkamat) [reply test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala:93]: Ok

**Review by girishkamat — APPROVED** [review]

**Pull request merged**